### DISCUSSION TITLE
v5: How to define the type of a client object?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mqttjs/MQTT.js#1725] [Q&A]
I am using `mqtt@5.1.3`. I have a variable-length array of MQTT clients.

```ts
import {connect, MqttClient} from 'mqtt'

export interface MqttClients {
  [key: string]: MqttClient
}

const clients = {
  a: connect(...),
  b: connect(...),
}

// TS2339: Property `subscribe` does not exist on type `OperatorFunction<unknown, unknown>` (on `subscribe`)
clients.a.subscribe('topic')

// 
TS2352: Conversion of type `OperatorFunction<unknown, unknown>` to type `typeof MqttClient` may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to `unknown` first. (on `clients.a as typeof MqttClient`)
(clients.a as typeof MqttClient).subscribe('asdf')

// TS1005: `?` expected. (on the first `)`)
(clients.a as typeof MqttClient extends TypedEventEmitter<MqttClientEventCallbacks>).subscribe('asdf', null, (err: any) => {
  if (err) {
    this.logger.error([`[${schema}] MQTT subscribe : Error :`, err], {each: true})
  }
})
```

It does not matter if I add `as unknown`, as I get these errors regardless.

I get these errors at least on `subscribe()`, `on()` and `end()` methods.

What is the correct way to define the client object? I’d prefer doing it as simply as `typeof MqttClient`, however, I’d be okay withany type definition.

Thanks in advance! :pray:

**Comment by robertsLando** [comment]
MqttClient type should be fine here, maybe the error is:

```
const clients: MqttClients = {
  a: connect(...),
  b: connect(...),
}
```

**Reply by tukusejssirs** [reply]
No, that is a problem. Actually, I only forgot to do that in this demo code, not in the actual code I use.

I get the same errors when I add the type too.

**Reply by robertsLando** [reply]
![2023-10-26_11-13](https://github.com/mqttjs/MQTT.js/assets/11502495/5b15a516-8b51-4823-b8e9-df5d8a50dc1a)
When I write the exact same code on my side typescript doesn't throw any error:
